src/allocator_defrag.c

```diff
@@ -0,0 +1,426 @@
+    assert(key_info->keylen <= sizeof(key_info->key) / sizeof(key_info->key[0]));
+    return res;
+}
+
+/* Query jemalloc control interface using previously extracted key (with jeQueryKeyInit) instead of name string.
+ * This interface (named MIB in jemalloc) is faster as it avoids string dict lookup at run-time. */
+static inline int jeQueryCtlInterface(const jeMallctlKey *key_info, void *value) {
+    size_t sz = sizeof(size_t);
+    return je_mallctlbymib(key_info->key, key_info->keylen, value, &sz, NULL, 0);
+}
+
+static inline int binQueryHelperInitialization(jeBinInfoKeys *helper, unsigned bin_index) {
+    char mallctl_name[128];
+
+    /* Mib of fetch number of used regions in the bin */
+    snprintf(mallctl_name, sizeof(mallctl_name), "stats.arenas." STRINGIFY(ARENA_TO_QUERY) ".bins.%d.curregs", bin_index);
+    if (jeQueryKeyInit(mallctl_name, &helper->curr_regs) != 0) return -1;
+    /* Mib of fetch number of current slabs in the bin */
+    snprintf(mallctl_name, sizeof(mallctl_name), "stats.arenas." STRINGIFY(ARENA_TO_QUERY) ".bins.%d.curslabs", bin_index);
+    if (jeQueryKeyInit(mallctl_name, &helper->curr_slabs) != 0) return -1;
+    /* Mib of fetch nonfull slabs */
+    snprintf(mallctl_name, sizeof(mallctl_name), "stats.arenas." STRINGIFY(ARENA_TO_QUERY) ".bins.%d.nonfull_slabs", bin_index);
+    if (jeQueryKeyInit(mallctl_name, &helper->nonfull_slabs) != 0) return -1;
+
+    return 0;
+}
+
+/* Initializes the defragmentation system for the jemalloc memory allocator.
+ *
+ * This function performs the necessary setup and initialization steps for the defragmentation system.
+ * It retrieves the configuration information for the jemalloc arenas and bins, and initializes the usage
+ * statistics data structure.
+ *
+ * return 0 on success, or a non-zero error code on failure.
+ *
+ * The initialization process involves the following steps:
+ * 1. Check if defragmentation is supported by the current jemalloc version.
+ * 2. Retrieve the arena bin configuration information using the `je_mallctlbymib` function.
+ * 3. Initialize the `usage_latest` structure with the bin usage statistics and configuration data.
+ * 4. Set the `defrag_supported` flag to indicate that defragmentation is enabled.
+ *
+ * Note: This function must be called before using any other defragmentation-related functionality.
+ * It should be called during the initialization phase of the code that uses the
+ * defragmentation feature.
+ */
+int allocatorDefragInit(void) {
+    char mallctl_name[100];
+    jeBinInfo *bin_info;
+    size_t sz;
+    int je_res;
+
+    /* the init should be called only once, fail if unexpected call */
+    assert(!defrag_supported);
+
+    /* Get the mib of the per memory pointers query command that is used during defrag scan over memory */
+    if (jeQueryKeyInit("experimental.utilization.batch_query", &je_cb.util_batch_query) != 0) return -1;
+
+    je_res = jeQueryKeyInit("epoch", &je_cb.epoch);
+    assert(je_res == 0);
```

Comment:
I don't think you understood my previous comment.  In the first call, on 267, you return -1 if it fails.  On the 2nd call, you assert.  Why are there 2 different patterns for the same function call?

Reply:
The first case, could fail and does not o dictate bug by itself, because the jemalloc version does not have the query api, from this code perspective it's a valid case and -1 is returned to indicate that init didn't succeed. The following calls are expected to succeed and are asserted to verify. I moved the call out of the assert following your comment I believe.